core/opendaq/modulemanager/src/module_manager_impl.cpp

```diff
@@ -950,32 +968,80 @@ ListPtr<IMirroredDeviceConfig> ModuleManagerImpl::getAllDevicesRecursively(const
     return result;
 }
 
-StringPtr ModuleManagerImpl::getPreferredStreamingAddress(const DevicePtr& device)
+AddressInfoPtr ModuleManagerImpl::getDeviceConnectionAddress(const DevicePtr& device)
 {
     const auto info = device.getInfo();
     const auto configConnection = info.getConfigurationConnectionInfo();
+    const auto deviceInfoConnectionString = info.getConnectionString();
+
     if (!configConnection.assigned())
-        return "";
+        return nullptr;
 
-    const auto addressInfo = configConnection.getAddressInfo();
-    if (addressInfo.assigned() && addressInfo.getCount() > 0)
-        return addressInfo[0].getAddress();
+    const auto deviceConnectionString =
+        (deviceInfoConnectionString.assigned() && deviceInfoConnectionString.getLength())
+            ? deviceInfoConnectionString
+            : configConnection.getConnectionString();
 
-    return "";
+    for (const auto& addressInfo : configConnection.getAddressInfo())
+    {
+        if (deviceConnectionString == addressInfo.getConnectionString())
+            return addressInfo;
+    }
+
+    return nullptr;
+}
+
+AddressInfoPtr ModuleManagerImpl::findStreamingAddress(const ListPtr<IAddressInfo>& availableAddresses,
+                                                       const AddressInfoPtr& deviceConnectionAddress,
+                                                       const StringPtr& primaryAddressType)
+{
+    if (isValidConnectionAddressType(primaryAddressType)) // restrict by connection address type
+    {
+        // Attempt to reuse the address of device connection if it meets type constraints
+        if (deviceConnectionAddress.assigned() && deviceConnectionAddress.getType() == primaryAddressType)
+        {
+            for (const auto& addressInfo : availableAddresses)
+                if (addressInfo.getAddress() == deviceConnectionAddress.getAddress())
+                    return addressInfo;
+        }
+
+        // If the device connection address is unavailable for streaming, search for any address matching type constraints
+        for (const auto& addressInfo : availableAddresses)
+        {
+            if (addressInfo.getType() == primaryAddressType)
+                return addressInfo;
+        }
+        LOG_W("Server streaming capability does not provide any addresses of primary {} type", primaryAddressType);
+    }
+
+    // Attempt to reuse the address of device connection
+    for (const auto& addressInfo : availableAddresses)
+    {
+        if (addressInfo.getAddress() == deviceConnectionAddress.getAddress())
+            return addressInfo;
+    }
+
+    return nullptr;
+}
+
+bool ModuleManagerImpl::isValidConnectionAddressType(const StringPtr& connectionAddressType)
+{
+    return (connectionAddressType == "IPv4" || connectionAddressType == "IPv6") ? true : false;
```

Comment:
return (connectionAddressType == "IPv4" || connectionAddressType == "IPv6") ? true : false;
```suggestion
    return connectionAddressType == "IPv4" || connectionAddressType == "IPv6"
```